### DISCUSSION TITLE
coredns support edns client subnet?

### DISCUSSION THREAD

**Discussion by 395096713** [coredns/coredns#6950] [Q&A] [answered]
hi admin:
i use coredns in my system for sometime, but i need use edns client subnet in certain scenarios, i search code, there is no plugin for this, i found plugin/pkg/edns mention some supported options，but no clientIP，so i want to know if support ecs？and forward can cache ecs answers？

**Comment by chrisohaver** [comment]
Can you describe more precisely what you want to do? _geoip_ and _rewrite_ plugins may be of interest depending on what you want to do.

**Reply by 395096713** [reply]
i have read the doc of geoip and rewrite, but It feels a bit different from my scene；Maybe I don't know enough about them; my scene is
1、use coredns as Cache layer DNS
2、use bind as DNS data source，open edns and views/acl for smartDns
3、so i need coredns cache the result get from bind according to  IP subsent
4、for example
use1 ip 10.10.10.10/24
use2 ip 20.20.20.20/24
bind view1/acl1  10.10.10.0/24
aaa.bbb.ccc A 10.1.1.1
bind view2/acl2 20.20.20.0/24
aaa.bbb.ccc A 20.2.2.2
when use1 request coredns，coredns use edns0 carryingIP 10.10.10.10/24 request bind；bind response answer aaa.bbb.ccc A 10.1.1.1 with edns0 subnet 10.10.10.0/24
when use2 request coredns，coredns use edns0 carryingIP 20.20.20.20/24 request bind；bind response answer aaa.bbb.ccc A 20.2.2.2 with edns0 subnet 20.20.20.0/24
different users can get different answers from the same coredns,coredns can cache answers use ecs(edns client subnet)

**Reply by chrisohaver** [reply]
apologies - I read this a few times and I don't understand it.

**Reply by chrisohaver** [reply]
1. I think this means you are using coredns as a caching forwarder?
2. I think this means that coredns forwards to a bind server.  (I dont know what "open edns and views/acl for smartDns" means)
3. I dont know what is meant by "according to IP subnet"

**Reply by chrisohaver** [reply]
OK, with help from @SuperQ, I think I understand the ask.  Thanks!

**Comment by johnbelamaric** [comment]
I think `metadata` + `metadata_edns0` + `view` (or `firewall` which is external and in the policy repo) solve that, don't they @chrisohaver ?

**Reply by johnbelamaric** [reply]
Oh, I see we never made `metadata_edns0` a default plugin, it is here: https://github.com/coredns/metadata_edns0

**Reply by chrisohaver** [reply]
Could be? But I don't understand what (numbers) wants to achieve, so I'm not sure.

**Comment by chrisohaver** [comment, marked as answer]
Does forward retain edns0 information when forwarding upstream?  Or does it strip it out.  I recall recurring disagreements on whether or not forward should retain edns0 from the request, but I dont recall which way we landed.  The discussions focused on whether or not forwarding DNS counts as a "hop," and if so the RFCs say that edns0 information should be retained "hop to hop" only.

If forward _does_ retain edns0, it could be done...

The requirement is to have a separate cache per subnet, and the only way to do that is to create a view for every possible subnet.  If the remote IP of the request matches the real ip of the client (aligned with the ecs data in the request), then you don't need `metadata` or `metadata_edns0`, since you can define the views to split by subnet based on the remote IP.  But if the requests are L3 routed such that the remote IP is no longer the  real ip of the client, then you'd need those plug-ins to retrieve the ECS values and base the views on those (@johnbelamaric's suggestion above).

Some years ago, there was a PR (rejected) to extend the cache plugin that allowed adding of metadata to cache hash/key.  The goal was accommodate spit DNS situations like this.  Such an option would enable a single cache solution that does not rely on views.  Applied to this ask, it would still require the `metadata_edns0` plugin however.

I dont recall if there were there objections to including the `metadata_edns0` in CoreDNS.  It seems like a simple utilitarian plugin.  I think we should include it.

**Reply by 395096713** [reply]
yes, The requirement is to have a separate cache per subnet, and answer diferent result according to client subnet;
I used unbound before, it support; search ecs   https://nlnetlabs.nl/documentation/unbound/unbound.conf/
so i want to know coredns support it ，i want use the same feature

**Reply by 395096713** [reply]
Supplement:
when i use unbound as cache layer，it's easy for open ecs support,just two cmd; 
client-subnet-zone: .
max-client-subnet-ipv4: 32
